subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,104 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	/* Current LTV offset contains the offset in the AD structure.
+	 * The length and type are included in this offset.
+	 */
+	uint16_t current_ltv_offset;
+};
+
+static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+static const struct bt_data *ad_stream_current_ltv_get(struct ad_stream *stream)
+{
+	const struct bt_data *current_ltv = &stream->ad[stream->ad_index].data[stream->data_index];
+	const bool current_ltv_complete = (stream->current_ltv_offset == current_ltv->data_len + 2);
+
+	if (current_ltv_complete) {
+		stream->current_ltv_offset = 0;
+
+		if (stream->data_index + 1 == stream->ad[stream->ad_index].len) {
+			stream->data_index = 0;
+			stream->ad_index++;
+		} else {
+			stream->data_index++;
+		}
+	}
+
+	if (ad_stream_is_empty(stream)) {
+		return NULL;
+	} else {
+		return &stream->ad[stream->ad_index].data[stream->data_index];
+	}
+}
+
+/** @brief Read at max buf_len data from the flattened AD stream.
+ *
+ * The read data can contain multiple LTV AD structures.
+ *
+ * @param stream  AD stream, @ref ad_stream_new
+ * @param buf     Buffer where the data will be put
+ * @param buf_len Buffer length
+ *
+ * @returns The written length
+ */
+static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t buf_len)
+{
+	uint8_t written_len = 0;
```

Comment:
Read instead of write?

```suggestion
 * @returns The read length
 */
static uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t buf_len)
{
	uint8_t read_len = 0;
```

Reply:
I guess that is a matter of semantics. I can change it if you want